### PR TITLE
Create automated tests for device filter

### PR DESCRIPTION
**Pull request by BiancaJoca**
ZS2022-34
Create automated tests for device filter

- Filter is working as expected
- Filter is working with multiple fields used
- Clear All button resets the filter fields

### CODE REVIEW

**Comment by DariusChis** [comment]
please solve conflicts

File: src/main/java/driver/DriverFactory.java [hunk 1]
```diff
@@ -20,15 +20,15 @@ public DriverFactory(WebDriver driver) {
...
     @FindBy(id = "password")
-    private WebElement passwordField;
+    private static WebElement passwordField;
 
     @FindBy(id = "log-in-button")
-    private WebElement signInButton;
+    private static WebElement signInButton;
 
-    public void signInAsSuperAdmin() {
+    public static void signInAsSuperAdmin() {
```
Comment (CaprarClaudiu) [inline src/main/java/driver/DriverFactory.java]: do we need to change these to static?

File: src/main/java/pageObjects/Filter.java [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+    @FindBy(id = "device-filter-asignee-autocomplete-option-0")
+    static WebElement assignedToOption;
+    @FindBy(id = "device-filter-status-autocomplete-option-0")
+    static WebElement filterStatusOption;
+    @FindBy(id = " device-filter-tags-autocomplete-option-0")
+    static WebElement filterTagsOption;
+    @FindBy(id = "device-filter-clearButton")
+    static WebElement clearAllButton;
+
+    public static void filterBySubcategory() {
```
Comment (CaprarClaudiu) [inline src/main/java/pageObjects/Filter.java]: do we need this to be static?

File: src/main/java/pageObjects/Filter.java [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+        filterButton.click();
+        wait.until(ExpectedConditions.visibilityOf(filterTags));
+        filterTags.click();
+        wait.until(ExpectedConditions.visibilityOf(filterTagsOption));
+        filterTagsOption.click();
+        wait.until(ExpectedConditions.visibilityOf(applyButton));
+        applyButton.click();
+        wait.until(ExpectedConditions.urlToBe(Constants.DEVICES_URL));
+    }
+    public static void filterBySubcategoryAndName() {
```
Comment (CaprarClaudiu) [inline src/main/java/pageObjects/Filter.java]: do we need this to be static?

File: src/main/java/pageObjects/Filter.java [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+        wait.until(ExpectedConditions.visibilityOf(laptopOption));
+        laptopOption.click();
+        wait.until(ExpectedConditions.visibilityOf(filterName));
+        filterName.click();
+        filterName.sendKeys("Nume");
+        wait.until(ExpectedConditions.visibilityOf(applyButton));
+        applyButton.click();
+        wait.until(ExpectedConditions.urlToBe(Constants.DEVICES_URL));
+    }
+    public static void clearAllButton() {
```
Comment (CaprarClaudiu) [inline src/main/java/pageObjects/Filter.java]: do we need this to be static? also the name is not very clear (pun intended). could you rename it?

File: src/test/java/pageObjects/FilterTest.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        filter.filterBySubcategory();
+        String table = driver.findElement(By.id("device-table")).getText();
+        Assert.assertTrue(table.contains("LAPTOP"));
+    }
+    @Test
+    public void filterByName() {
+        filter.filterByName();
+        String table = driver.findElement(By.id("device-table")).getText();
+        Assert.assertTrue(table.contains("Name"));
+    }
```
Comment (CaprarClaudiu) [inline src/test/java/pageObjects/FilterTest.java]: correct me if im wrong but if we get the text from the whole table, we also get the header which has a column called "Name". this means we will always have this test pass even if the filter is not working as long as the device table loads

File: src/test/java/pageObjects/FilterTest.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        filter.filterBySubcategory();
+        String table = driver.findElement(By.id("device-table")).getText();
+        Assert.assertTrue(table.contains("Assigned"));
+    }
+    @Test
+    public void filterByTags() {
+        filter.filterBySubcategory();
+        String table = driver.findElement(By.id("device-table")).getText();
+        Assert.assertTrue(table.contains("pls_dnt_dell_test"));
+    }
```
Comment (CaprarClaudiu) [inline src/test/java/pageObjects/FilterTest.java]: are you sure you are calling the correct method for your test? :)

File: src/test/java/pageObjects/FilterTest.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+        String table = driver.findElement(By.id("device-table")).getText();
+        Assert.assertTrue(table.contains("LAPTOP"));
+        Assert.assertTrue(table.contains("Name"));
+    }
+    @Test
+    public void clearAllButton() {
+        filter.clearAllButton();
+        String table = driver.findElement(By.id("device-table")).getText();
+        String URL = driver.getCurrentUrl();
+        Assert.assertEquals(URL, Constants.DEVICES_URL);
```
Comment (CaprarClaudiu) [inline src/test/java/pageObjects/FilterTest.java]: here you have the unused string table. you can either delete it or add a new assert. i suggest that you should rethink this test. try something like this: open filter -> filter by something -> wait for the page to load -> open filter and click clear button -> then open filter again and check if the field is be empty.

**Review by CaprarClaudiu — COMMENTED** [review]
Resolve conflicts

File: src/main/java/pageObjects/Filter.java [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+    @FindBy(id = "device-filter-status-autocomplete")
+    WebElement filterStatus;
+    @FindBy(id = "device-filter-tags-autocomplete")
+    WebElement filterTags;
+    @FindBy(id = "device-filter-invoiceDate")
+    WebElement filterInvoiceDate;
+    @FindBy(id = "device-filter-apply-button")
+     WebElement applyButton;
+    @FindBy(id = "")
+    WebElement option0;
```
Comment (CaprarClaudiu) [inline src/main/java/pageObjects/Filter.java]: are we using this webelement? if we do i dont think that is the correct ID

File: src/main/java/pageObjects/Filter.java [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+     WebElement laptopOption;
+    @FindBy(id = "device-filter-asignee-autocomplete-option-0")
+     WebElement assignedToOption;
+    @FindBy(id = "device-filter-status-autocomplete-option-0")
+     WebElement filterStatusOption;
+    @FindBy(id = "device-filter-tags-autocomplete-option-0")
+     WebElement filterTagsOption;
+    @FindBy(id = "device-filter-clearButton")
+     WebElement clearAllButton;
+    String name = "0";
```
Comment (CaprarClaudiu) [inline src/main/java/pageObjects/Filter.java]: i think this should be placed in the constants file

File: src/test/java/pageObjects/FilterTest.java [hunk 10]
```diff
@@ -0,0 +1,64 @@
+package pageObjects;
+
+import org.openqa.selenium.By;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+import utils.Constants;
```
Comment (CaprarClaudiu) [inline src/test/java/pageObjects/FilterTest.java:6]: :)

**Review by CaprarClaudiu — CHANGES_REQUESTED** [review]

**Review by CaprarClaudiu — APPROVED** [review]

**Review by ScorusMihai — APPROVED** [review]

**Pull request merged**